idem_azurerm/exec/azurerm/application_insights/component.py

```diff
@@ -0,0 +1,299 @@
+# -*- coding: utf-8 -*-
+"""
+Azure Resource Manager (ARM) Application Insights Component Operations Execution Module
+
+.. versionadded:: 3.0.0
+
+:maintainer: <devops@eitr.tech>
+:configuration: This module requires Azure Resource Manager credentials to be passed as keyword arguments
+    to every function or via acct in order to work properly.
+
+    Required provider parameters:
+
+    if using username and password:
+      * ``subscription_id``
+      * ``username``
+      * ``password``
+
+    if using a service principal:
+      * ``subscription_id``
+      * ``tenant``
+      * ``client_id``
+      * ``secret``
+
+    Optional provider parameters:
+
+**cloud_environment**: Used to point the cloud driver to different API endpoints, such as Azure GovCloud.
+    Possible values:
+      * ``AZURE_PUBLIC_CLOUD`` (default)
+      * ``AZURE_CHINA_CLOUD``
+      * ``AZURE_US_GOV_CLOUD``
+      * ``AZURE_GERMAN_CLOUD``
+
+"""
+# Python libs
+from __future__ import absolute_import
+import logging
+
+# Azure libs
+HAS_LIBS = False
+try:
+    import azure.mgmt.web.models  # pylint: disable=unused-import
+    from msrestazure.azure_exceptions import CloudError
+    from msrest.exceptions import ValidationError
+
+    HAS_LIBS = True
+except ImportError:
+    pass
+
+__func_alias__ = {"list_": "list"}
+
+log = logging.getLogger(__name__)
+
+
+async def create_or_update(
+    hub,
+    ctx,
+    name,
+    resource_group,
+    kind,
+    application_type,
+    retention=None,
+    immediate_purge_data=None,
+    disable_ip_masking=None,
+    ingestion_public_network_access=None,
+    query_public_network_access=None,
+    tags=None,
+    **kwargs,
+):
+    """
+    .. versionadded:: 3.0.0
+
+    Creates (or updates) an Application Insights component.
+
+    :param name: The name of the Application Insights component resource.
+
+    :param resource_group: The name of the resource group.
+
+    :param kind: The kind of application that this component refers to, used to customize UI. This value is a freeform
+        string, values should typically be one of the following: "web", "ios", "other", "store", "java", "phone".
+
+    :param application_type: Type of application being monitored. Possible values include: 'web', 'other'.
+        Default to "web".
+
+    :param retention: Retention period in days. Defaults to 90.
+
+    :param immediate_purge_data: A boolean value representing whether or not data should be purged immediately after
+        30 days.
+
+    :param disable_ip_masking: Disable IP masking.
+
+    :param tags: A dictionary of strings can be passed as tag metadata to the Application Insights Component object.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        azurerm.application_insights.component.create_or_update test_name test_group test_kind test_type
```

Comment:
if `kind` and `application_type` have specific valid values, let's use those in the examples.